phac-nml/irida-next
# Sample Exports: Backend - Attachment Format Filtering

## What does this PR do and why?
This PR adds the backend logic for sample export format filtering. This will allow users to create export for samples which only contain the formats they're interested in.

## Screenshots or screen recordings
_Screenshots are required for UI changes, and strongly recommended for all other pull requests._

## How to set up and validate locally
1. Reseed if needed with `bin/rails db:drop/create/migrate/seed`
In `rails c`:
2. Assign:
```
user = User.find_by(email: 'user7@email.com')
params_1 = {'export_type' => 'sample', 'export_parameters' => {'ids' => [Sample.find_by(name: 'Yersinia pestis/Outbreak 2022 Sample 10').id, Sample.find_by(name: 'Yersinia pestis/Outbreak 2022 Sample 9').id], 'namespace_id' => Sample.find_by(name: 'Yersinia pestis/Outbreak 2022 Sample 10').project.namespace.id, 'attachment_formats' => ['fastq']},  'name' => 'test export'}
```
3. Create the export:
```
DataExports::CreateService.new(user, params_1).execute
```
4. After 30s, download the export by copying and pasting the given URL into a browser. Replace `localhost` with `localhost:3000` if needed:
```
Rails.application.routes.url_helpers.rails_blob_url(DataExport.last.file)
```
5. Ensure the sample attachment folders only contain the `fastq` files and not the `fasta` files.
6. Ensure the `manifest.json` only contains the `fastq` files and not the `fasta` files.

7. Repeat the above with the params:
```
params_1 = {'export_type' => 'sample', 'export_parameters' => {'ids' => [Sample.find_by(name: 'Yersinia pestis/Outbreak 2022 Sample 10').id, Sample.find_by(name: 'Yersinia pestis/Outbreak 2022 Sample 9').id], 'namespace_id' => Sample.find_by(name: 'Yersinia pestis/Outbreak 2022 Sample 10').project.namespace.id, 'attachment_formats' => ['fasta']},  'name' => 'test export'}
```
- This should only have `fasta` files.

8. Repeat the above with the params:
```
params_1 = {'export_type' => 'sample', 'export_parameters' => {'ids' => [Sample.find_by(name: 'Yersinia pestis/Outbreak 2022 Sample 10').id, Sample.find_by(name: 'Yersinia pestis/Outbreak 2022 Sample 9').id], 'namespace_id' => Sample.find_by(name: 'Yersinia pestis/Outbreak 2022 Sample 10').project.namespace.id, 'attachment_formats' => ['text']},  'name' => 'test export'}
```
- This should contain no files other than the `manifest.json`.

## PR acceptance checklist
This checklist encourages us to confirm any changes have been analyzed to reduce risks in quality, performance, reliability, security, and maintainability.
